src/normalization.rs

```diff
@@ -0,0 +1,163 @@
+//! Normalizer trait and default implementations.
+//!
+//! Normalization is how strings are normalized into characters you want to
+//! consider equal. For example, if you want a case insensitive comparison,
+//! you might consider the [LowerCaseNormalizer].
+//!
+//! ```
+//! # use fuzzywuzzy::normalization::{Normalizer, LowerCaseNormalizer, FormCNormalization, MultipleNormalizer};
+//! assert_eq!(LowerCaseNormalizer.normalize("this STRING"), LowerCaseNormalizer.normalize("THIS string"));
+//! let a1 = "ä"; // U+00E4
+//! let a2 = "ä"; // U+0061 + U+0308
+//! let a3 = "Ä"; // U+0041 + U+0308
+//! assert_ne!(a1, a2);
+//! assert_eq!(FormCNormalization.normalize(a2), a1);
+//! let multiple_normalizers = MultipleNormalizer::with(vec![Box::new(LowerCaseNormalizer), Box::new(FormCNormalization)]);
+//! assert_eq!(multiple_normalizers.normalize(a3), a1);
+//! ```
+
+/// Represents a strategy for normalizing string characters into a canonical value of their equivalence class.
+///
+/// i.e., in a case-insensitive context, 'a' might be the canonical value for the equivalence class of ASCII A's: `['a', 'A']`.
+pub trait Normalizer {
+    fn normalize(&self, s: &str) -> String;
+}
+
+impl<F: Fn(&str) -> String> Normalizer for F {
+    fn normalize(&self, s: &str) -> String {
+        self(s)
+    }
+}
+
+/// Doesn't modify any characters. The Identity-transform [Normalizer].
+pub struct PassthroughNormalizer;
+
+impl Normalizer for PassthroughNormalizer {
+    fn normalize(&self, s: &str) -> String {
+        s.into()
+    }
+}
+
+// ew, need a better name
+/// Compose a sequence of [Normalizer]s together into one [Normalizer].
+///
+/// They are executed in order.
+pub struct MultipleNormalizer {
+    normalizers: Vec<Box<dyn Normalizer>>,
+}
+
+impl MultipleNormalizer {
+    pub fn with(normalizers: Vec<Box<dyn Normalizer>>) -> MultipleNormalizer {
+        MultipleNormalizer { normalizers }
+    }
+}
+
+impl Normalizer for MultipleNormalizer {
+    fn normalize(&self, s: &str) -> String {
+        let mut current = s.to_owned();
+        for normalizer in self.normalizers.iter() {
+            current = normalizer.normalize(&current);
+        }
+        current
+    }
+}
+
+/// Normalizes strings by lower-casing all letters.
+pub struct LowerCaseNormalizer;
+
+impl Normalizer for LowerCaseNormalizer {
+    fn normalize(&self, s: &str) -> String {
+        s.to_lowercase()
+    }
+}
+
+/// Removes non-ASCII codepoints.
+pub struct AsciiOnlyFilter;
```

Comment:
Is there any distinction between a "filter" and a "normalizer"?

Reply:
Type-wise, no. Semantically... sort of, but still kind of not really.

Reply:
Thoughts on renaming to `AsciiNormalizer` or `AsciiOnlyNormalizer`?